Allow for specifying device with element numbers

nkanazawa1989 opened the pull request:
<!--
⚠️ If you do not respect this template, your pull request will be closed.
⚠️ Your pull request title should be short detailed and understandable for all.
⚠️ Also, please add it in the CHANGELOG file under Unreleased section.
⚠️ If your pull request fixes an open issue, please link to the issue.

✅ I have added the tests to cover my changes.
✅ I have updated the documentation accordingly.
✅ I have read the CONTRIBUTING document.
-->

### Summary
close #2479 

This PR allows for specifying device with numbers of `n_drive`, `n_control`, `n_measure`, and `n_register`. As discussed in #2479, The class name `DeviceSpecification` is renamed to `PulseSpecification` because this class consists of a part of device information (mainly related to topology of device).

### Details and comments

- #qubit != `n_drive` or `n_control`
e.g. allowing for a qubit with XY control and Z control, multiple control channels, control channels to tunable coupler, etc...

- #acquire, #mem_slots = `n_measure`
i.e. these measurement chain is usually bound. This may be relevant to #2524 .

- `n_registers` != `n_measure`
e.g. measured results can be loaded to arbitrary registers and classical operations can be applied to each register.

All channels can be directly accessed without using qubit (`device.q[0].drive` and `device.d[0]` are both possible, and they refer the same object). These channels will be assigned to qubit/tunable coupler/... by using device Hamiltonian.

taalexander commented on qiskit/pulse/channels/pulse_specification.py:
```diff
@@ -0,0 +1,139 @@
+                 n_measure: int,
+                 n_registers: int,
+                 buffer: int = 0,
+                 hamiltonian=None):
```
This should be added as a separate PR along with all other hamiltonian related information.

taalexander commented on qiskit/pulse/channels/pulse_specification.py:
```diff
@@ -0,0 +1,139 @@
+
+
+class PulseSpecification:
+    """Implement a device specification."""
```
Could you please provide a more detailed summary and description of the purpose/use case of this class?

taalexander commented on qiskit/pulse/channels/pulse_specification.py:
```diff
@@ -0,0 +1,139 @@
+        defaults = backend.defaults()
+
+        # TODO: define n_drive and n_acquire explicitly in backend
+        n_drive = len(configuration.qubit_lo_range)
```
This should continue to be n_qubits, not qubit_lo_range which is guaranteed by the spec to be of length `n_qubits`. Same for measurement information. I believe you are changing this due to a future change in the spec, but this should be done as multiple steps.

taalexander commented on qiskit/pulse/channels/pulse_specification.py:
```diff
@@ -0,0 +1,139 @@
+                                  n_measure=n_measure, n_registers=n_registers,
+                                  buffer=buffer, hamiltonian=hamiltonian)
+
+    def __eq__(self, other):
```
I'm not really sure of the use case for checking equality of two device specification? This might be adding too many features. It might be better to re-add at a later date if necessary.

nkanazawa1989 replied:
If we have some devices with the same topology, we don't need to recreate schedule. If we want to execute a schedule in parallel with multiple devices, we may want to check if the devices have the same topology. But I think such situation doesn't happen soon and we can remove equality check.

taalexander commented on qiskit/pulse/channels/pulse_specification.py:
```diff
@@ -0,0 +1,139 @@
+        return False
+
+    @property
+    def d(self) -> List[DriveChannel]:
```
Could these be something like `drives`, `memoryslots`, `qubits` etc? I think these names are too short.

taalexander commented on test/python/compiler/test_assembler.py:
```diff
@@ -262,7 +262,7 @@ class TestPulseAssembler(QiskitTestCase):
     """Tests for assembling schedules to qobj."""
```
The tests should be duplicated for now for `DeviceSpecification` until deprecation, although tests that use `DeviceSpecification` can just be switched to `PulseSpecification` as you have.

nkanazawa1989 commented:
@taalexander thank you for reviewing this. I updated the PR to align with your comments. unittests for `PulseSpecification` use `device.drives[0]` rather than `device.qubits[0].drive` because the later style implicitly includes qubit-channel mapping from device Hamiltonian. This should be tested separately in the new module `DeviceTopology` (I added this module and unittest for future extendability for device Hamiltonian).

Device Hamiltonian and n_drives/n_measures will be updated in following PRs after the spec is updated.

nkanazawa1989 commented on qiskit/pulse/channels/pulse_specification.py:
```diff
@@ -0,0 +1,142 @@
+from .device_topology import DeviceTopology
+
+
+class PulseSpecification:
```
That's good suggestion. I like `PulseChannelSpec`.

nkanazawa1989 commented:
@taalexander thank you for suggestions! the PR is updated.

nkanazawa1989 commented on qiskit/pulse/channels/pulse_channel_spec.py:
```diff
@@ -0,0 +1,155 @@
+                                        self.measures, self.acquires)
+
+    @classmethod
+    def from_device(cls, backend):
```
Do we change this to `from_system` ?

taalexander replied:
Yes, but only in the new pulse channel spec and not the old DeviceSpecification.

taalexander commented on qiskit/pulse/channels/pulse_channel_spec.py:
```diff
@@ -0,0 +1,155 @@
+    """A helper class to support assembling channel objects and their mapping to qubits.
+    This class can be initialized with two methods shown as below.
+
+    1. When you have BaseBackend object of the target quantum device.
```
```suggestion
    1. With the `BaseBackend` object of the target pulse backend:
```

taalexander commented on qiskit/pulse/channels/pulse_channel_spec.py:
````diff
@@ -0,0 +1,155 @@
+        system = PulseChannelSpec.from_device(backend)
+        ```
+
+    2. When you know the number of elements constituting the target quantum device.
````
```suggestion
    2. By specifying the number of pulse elements constituting the target quantum system:
```

nkanazawa1989 replied:
I have a concern about using [quantum system](https://en.wikipedia.org/wiki/Quantum_system) to denote a system consisting of quantum chip + cryosystem + rte. This may remind us of a term in quantum physics. I think _quantum computing system_ sounds more suitable.

taalexander replied:
yes, I agree. Your suggestion makes sense.

nkanazawa1989 replied:
thanks, I updated.

nkanazawa1989 commented:
@taalexander thank you! I think every ~`device` is replaced by ~`system`.

taalexander commented on qiskit/pulse/channels/pulse_channel_spec.py:
```diff
@@ -91,9 +92,9 @@ def __init__(self,
 
     @classmethod
-    def from_device(cls, backend):
+    def from_system(cls, backend):
```
Do you think we should name this `from_backend` as this is what the method is doing?

nkanazawa1989 replied:
that's sounds reasonable.
